### PR TITLE
feat(eslint-plugin): add new 'strict' config

### LINKED ISSUE #4648: Create a "strict" config between "recommended-requiring-type-checking" and "all"
## Overview

Moving discussion out of https://twitter.com/JoshuaKGoldberg/status/1500610010576994306: it seems to me that there are at least two classifications of lint users we should target:

* **Starter**: most users at least to begin with. These users likely don't have a "power user" w.r.t. typescript-eslint on their project and thus don't want particularly strict, annoying options enabled. They should be given rules that are clearly demonstrate linting value.
    * Lint configuration target today: `recommended` or `recommend-requiring-type-checking`
* **Power**: users who have seen the power of typed linting and want to get the most out of it. They want our somewhat-opinionated strict configs that enforce best practices too -- and can handle the sometimes aggressive lint errors as a result.
    * Lint configuration target today: missing

I propose we create a `plugin:@typescript-eslint/strict` config that adds a suite of best practice rules on top of `recommended-requiring-type-checking`.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4648
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Adds a new `'strict'` level for the `recommended` metadata for rules. It indicates to include the rule in a new `plugin:@typescript-eslint/strict` config.

~Putting this PR up now as a draft as reference for #4648 before fleshing out docs.~ ✅

### CODE REVIEW

File: packages/eslint-plugin/src/configs/strict.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    '@typescript-eslint/consistent-type-definitions': 'warn',
+    'dot-notation': 'off',
+    '@typescript-eslint/dot-notation': 'warn',
+    '@typescript-eslint/no-base-to-string': 'warn',
+    '@typescript-eslint/no-confusing-non-null-assertion': 'warn',
+    '@typescript-eslint/no-dynamic-delete': 'warn',
+    '@typescript-eslint/no-extraneous-class': 'warn',
+    '@typescript-eslint/no-invalid-void-type': 'warn',
+    '@typescript-eslint/no-meaningless-void-operator': 'warn',
+    '@typescript-eslint/no-non-null-asserted-nullish-coalescing': 'warn',
```
Comment (bradzacher) [inline packages/eslint-plugin/src/configs/strict.ts:27]: wait - this isn't recommended?!?

Oh - I see - it was added after the last major!

**Review by bradzacher — APPROVED** [review]

File: docs/linting/PRESETS.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+---
```
[the remark this answers is not in the document]
Comment (armano2) [reply docs/linting/PRESETS.md:1]: we should def merge it, its way to confusing to read trough multiple files that describe same thing

File: packages/eslint-plugin/src/configs/strict.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    '@typescript-eslint/no-unnecessary-type-arguments': 'warn',
+    'no-useless-constructor': 'off',
+    '@typescript-eslint/no-useless-constructor': 'warn',
+    '@typescript-eslint/non-nullable-type-assertion-style': 'warn',
+    '@typescript-eslint/prefer-for-of': 'warn',
+    '@typescript-eslint/prefer-function-type': 'warn',
+    '@typescript-eslint/prefer-includes': 'warn',
+    '@typescript-eslint/prefer-literal-enum-member': 'warn',
+    '@typescript-eslint/prefer-nullish-coalescing': 'warn',
+    '@typescript-eslint/prefer-optional-chain': 'warn',
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/configs/strict.ts:41]: Agreed. I've felt the same pain and love-hate relationship. #4820

**Review by bradzacher — DISMISSED** [review]
LGTM!

File: packages/eslint-plugin/docs/rules/TEMPLATE.md [hunk 4]
```diff
@@ -52,6 +54,8 @@ For example if this rule requires a feature released in a certain TS version.
 
 ## Attributes
 
-- [ ] ✅ Recommended
+- Configs:
+  - [ ] ✅ Recommended
+  - [ ] ✔ Strict
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/TEMPLATE.md:59]: @JamesHenry suggested 🔒 instead of ✔️ and I love it. Will change.

**Review by bradzacher — APPROVED** [review]

**Comment by bradzacher** [comment]
feel free to admin merge once you've got this finalised - i'm happy with what you've got!

**Comment by JoshuaKGoldberg** [comment]
My main hesitation against pushing this is that we don't have a blog to nicely announce it on yet (#4957) 😂. But that's not a good reason. We can always posthumously write a blog post for this PR.

**Pull request merged**